[Design] 온보딩 UI 구현

0Hooni opened the pull request:
## 🔥 PR 타입 (하나 이상 선택)
- [x] 기능 추가
- [ ] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [ ] 기타 사소한 수정

## 🔗 연결된 이슈
- Connected: #47 
- [x] 이슈 해결 여부

## 📄 작업 내용
<!-- 작업한 내용을 두괄식으로 작성해주세요 -->
|    구현 내용    |   15 pro   |
| :-------------: | :----------: |
| 로그인 | <video src = "https://github.com/Honzapda/Honzapda_iOS/assets/37678646/5950c449-1cbe-4113-99f8-c3c2eaffacbd" width ="300"> |
| 회원가입 | <video src = "https://github.com/Honzapda/Honzapda_iOS/assets/37678646/c0bfc7b3-eea8-403d-a58c-da9889ee6ff4" width ="300"> |

## 💻 주요 코드 설명
`SignUpCurationView`
- 회원가입 후 보여지는 큐레이션 화면
- WrappingHStack 라이브러리 이용
- GrayOutlineButton은 View+ 에 구현된 항목
```swift
WrappingHStack(signUpCurationViewModel.signUpCurationModel,
               id: \.self, alignment: .center) { model in
    GrayOutlineButton(title: model.title, isSelected: model.isSelected)
}
```

## 👀 기타 더 이야기해볼 점
- 이번 PR 리뷰해주실 때 MVVM의 적용이 타당했는지와 타당했다면 올바르게 설계됐는지에 대해서 좀 알고싶어서 이부분 봐주시면 감사하겠습니다.
- 하나의 View 설계를 위해 라이브러리를 사용하는것에 대한 의견 공유 부탁드립니다.
- 이번 PR 라인이 긴 이유는 약관 동의 내용이 코드 안에 다 들어가서 그럽니다(암튼 그럼)

0Hooni committed:
chore: 기존 코드 간소화 및 고정사이즈 해제

- 기존에 frame으로 fixed한 사이즈를 사용하는 일부 View들을 화면 기준 Padding을 이용하여 재설계
- 일부 컴포넌트는 재사용 View로 교체

0Hooni committed:
refactor: AppleLogin SwiftUI로 전환

- UIKit을 이용하여 구현됐던 기존 애플로그인을 SwiftUI로 재설계

0Hooni committed:
add: Assets 추가

- 약관 동의 View 구현을 위한 Assets 추가

0Hooni committed:
refactor: UI 구성방식 변경

- ZStack을 사용하는 배경 처리 방식을 .background를 이용하여 구성된 컴포넌트 만큼만 사용하게 변경
- 일부 fixed size를 사용하는 부분을 padding 처리로 자연스럽게 위치하도록 변경

0Hooni committed:
desing: 약관동의 UI 구성

- MVVM 패턴 적용

0Hooni committed:
fix: #47 - 디렉토리 분류 추가

- View 폴더를 Main, Sub 두가지로 분류

0Hooni committed:
feat: #47 - 약관 동의 버튼 구현

- 약관 버튼 클릭시 popup
- popup 내부의 동의 버튼 클릭 시 약관 버튼 색상 활성화

0Hooni committed:
add: #47 - 에셋 추가

- 회원가입 배경화면 추가
- 회원가입 완료 배경화면 추가
- 회원가입 완료 아이콘 추가

0Hooni committed:
design: 커스텀 네비게이션 백버튼 추가

- color, title, dismiss 액션을 받아서 넣어주면 됨
- 기존에 지속적으로 커스텀 버튼 변수를 만들어줘야되는 부분을 개선하기 위해 추가

0Hooni committed:
design: 약관동의 및 회원가입 화면 구현

- 제곧내
- 회원가입 타이핑된 내용 체크에 대한 부분은 차후 API 추가시 같이 추가할 예정
- MVVM 패턴이 올바르게 적용되었는가에 대한 의문점이 있었음

0Hooni committed:
fix: 디렉토리 옮김

- SignUpCompleteView가 ViewModel 폴더에 있던 부분을 수정(View/Main으로 이동)

0Hooni committed:
feat: #47 - 회원가입 후 큐레이션 이동

- DispatchQueue를 이용하여 View 이동 구현

0Hooni committed:
chore: #47 - View extension 재정렬

- Global, extension, Private 순으로 들어가도록 설정

0Hooni committed:
add: #47 - wrappinghstack 라이브러리 추가

- Tag layout 구현을 위한 라이브러리 추가

0Hooni committed:
design: #47 - 회원가입 큐레이션 뷰 구현

- WrappedHStack 라이브러리 이용

0Hooni committed:
add: #47 - 주석 추가

- 구현 내용에 주석 추가

0Hooni committed:
Merge branch 'develop' into design/#47-onboarding-ui

youz2me commented on Honzapda/Global/Extensions/View+.swift:
```diff
@@ -51,6 +51,93 @@ struct HighlightedText: View {
+            dismiss()
+        } label: {
+            HStack {
+                Image(.iconChevronLeft) // '<' Image
```
오 이거 시스템 이미지 사용한건가요 ?? 시스템 이미지도 이렇게 쓸 수 있군용 ... 좋은 거 알아갑니다

0Hooni replied:
@youz2me 저도 최근에 알았습니다😂

youz2me commented on Honzapda/Global/Extensions/View+.swift:
```diff
@@ -51,6 +51,93 @@ struct HighlightedText: View {
+}
+
+struct GrayOutlineButton: View {
+    let title: String
+    @State var isSelected = false
+    
+    var body: some View {
+        if isSelected {
+            Text(title)
+                .font(.sCoreDream(.medium, size: 16))
+                .foregroundColor(.white)
+                .padding(EdgeInsets(top: 8, leading: 24, bottom: 8, trailing: 24))
+                .background(.primary05)
+                .overlay(
+                    RoundedRectangle(cornerRadius: 24)
+                        .inset(by: 0.5)
+                        .stroke(.primary05, lineWidth: 1)
+                    
+                )
+                .clipShape(RoundedRectangle(cornerRadius: 24))
+                .onTapGesture {
+                    isSelected = false
+                }
+        } else {
+            Text(title)
+                .font(.sCoreDream(.medium, size: 16))
+                .foregroundColor(.gray06)
+                .padding(EdgeInsets(top: 8, leading: 24, bottom: 8, trailing: 24))
+                .overlay(
+                    RoundedRectangle(cornerRadius: 24)
+                        .inset(by: 0.5)
+                        .stroke(.gray04, lineWidth: 1)
+                )
+                .frame(height: 40)
+                .onTapGesture {
+                    isSelected = true
+                }
+        }
+    }
+}
```
음? 근데 이런 것들은 익스텐션 말고 다른 파일로 분류해주는게 더 나아보입니다! 익스텐션에는 익스텐션만 넣는게 좋아보여용

0Hooni replied:
@youz2me 별도 파일로 반영하겠습니다

youz2me commented on Honzapda/Presentation/Onboarding/View/Main/SignUpCompleteView.swift:
```diff
@@ -0,0 +1,56 @@
+// MARK: - 회원가입 완료 View
+struct SignUpCompleteView: View {
+    // MARK: PARAMS
+    @State private var isTimerFinish = false
```
이것도 뷰모델로 빼는게 좋아보이빈당

0Hooni replied:
@youz2me 해당 변수 하나를 위해 모델과 뷰모델로 만드는것이 맞나싶어서 뺐었는데 짐깅은 이걸 뷰모델로 빼는게 좋다 생각한 이유가 있을까요?

youz2me commented on Honzapda/Presentation/Onboarding/View/Main/SignUpEmailTypeView copy.swift:
```diff
@@ -0,0 +1,78 @@
+    // MARK: PARAMS
+    @Environment(\.dismiss) private var dismiss
+    @ObservedObject var signUpViewModel = SignUpViewModel()
+    @State private var inputText = ""
+    let placeholder = "honzapda@gmail.com"
+    @State private var isNextButtonClicked = false
```
요것들도 뷰모델로 빼는게 어떨까요 ??

youz2me replied:
private도 일괄적용하면 좋을 것 같아요

youz2me commented on Honzapda/Presentation/Onboarding/View/Main/SignUpPasswordCheckTypeView.swift:
```diff
@@ -0,0 +1,77 @@
+    // MARK: PARAMS
+    @Environment(\.dismiss) private var dismiss
+    @ObservedObject var signUpViewModel = SignUpViewModel()
+    @State private var inputText = ""
+    let placeholder = "비밀번호 확인"
+    @State private var isNextButtonClicked = false
```
요... 친구들도 뷰모델로 빼는게 좋을 것 같습니다

0Hooni replied:
@youz2me 이부분 ViewModel로 만들면 EmailType, PasswordType, PasswordCheckType 총 세곳에서 하나의 VM을 끌고가면서 세개의 View에서 파라미터를 모두 받고, 이후 VM에서 세가지를 체크하며 서버로 보내는 비지니스 로직을 고려했습니다!☺️

혹시 이러한 설계 의도를 고려했을때 해당 변수들을 VM에 넣는다면 View가 다음 단계로 넘어갈 때 inputText와 placeholder, buttonClicked를 각 View에 맞게 다시 넣어주어야 되는데 혹시 좋은 방식이 있을까요?

youz2me commented on Honzapda/Presentation/Onboarding/View/Main/TermView.swift:
```diff
@@ -0,0 +1,167 @@
+            .navigationBarItems(leading: Button(
+                action: {dismiss()},
+                label: {
+                    Image(systemName: "xmark")
```
음? 여기서는 그냥 이름을 쓰셨네요 ,,, .iconXmark 또는 systemName: "xmark" 중 하나로 통일하는 게 좋아보입니다!

0Hooni replied:
@youz2me 이거는 확인해보니까 xmark가 .iconXmark로 불러와지지 않는것 같더라고여... ㅠㅠ 그래서 이렇게 했었습니다

youz2me commented on Honzapda/Presentation/Onboarding/View/Main/TermView.swift:
```diff
@@ -0,0 +1,167 @@
+private struct termDetailView: View {
+    // MARK: PARAMS
+    var termViewModel: TermViewModel
+    @Binding var isTermButtonClicked: Bool
+    
+    // MARK: BODY
+    var body: some View {
+        let curIndex = termViewModel.getCurrentTermIndex()
+        let termDetailTitle = termViewModel.terms[curIndex].title.replacingOccurrences(of: "\n", with: "")
+        let termDetailContent = termViewModel.terms[curIndex].content
```
저는 변수 선언하는 것들은 웬만하면 뷰모델로 빼려고 하고 있긴 합니다 ...! 뷰에서는 뷰의 역할만 할 수 있게 하는 게 더 좋은 것 같아요

youz2me reviewed (approved):
고생하셨어요 ~~ 뷰모델 관련 내용만 조금 개선하면 좋을 것 같습니다!

0Hooni committed:
fix: #48 - PR 코멘트 반영

- View extension과 내부 struct 분리
- 일부 ViewModel 밖으로 나와있던 변수를 ViewModel로 합침
- 잘못 작성된 파일명 수정
    - 회원가입(이메일 입력 View 이름)
    - 커스텀 네비게이션 백 버튼 네이밍 수정
